### PR TITLE
add struct_mapping_extend

### PR DESCRIPTION
**Pull request by waltcow**

### CODE REVIEW

File: builder/struct.go [hunk 1]
```diff
@@ -40,6 +40,33 @@ func (*Struct) Build(gen Generator, ctx *MethodContext, sourceID *xtype.JenID, s
 		}
 
 		targetFieldType := xtype.TypeOf(targetField.Type())
+
+		if mapFnName, ok := ctx.ExtendMapping[targetField.Name()]; ok {
+			if sourceMatch, err := source.StructField(targetField.Name(), ctx.MatchIgnoreCase, ctx.IgnoredFields); err == nil {
```
Comment (jmattheis) [inline builder/struct.go:45]: As described in https://github.com/jmattheis/goverter/issues/12#issuecomment-1010274080 let's start with only implementing the method without parameters.

> The first step would be to add goverter:mapExtend with support for only defining a method without arguments and one return value, which must be the same as the return type of the parameter that is defined inside goverter:mapExtend. The optional passing of the source value as first parameter should be done later.

File: scenario/3_struct_mapping_extend.yml [hunk 2]
```diff
@@ -0,0 +1,45 @@
+input:
+    input.go: |
+        package structs
+
+        // goverter:converter
+        // goverter:mapExtend PluckAgeInt
+        // goverter:mapExtend PlusScoreInt
```
Comment (jmattheis) [inline scenario/3_struct_mapping_extend.yml:7]: This shouldn't be needed, defining it inside on the converter method should be enough.

File: generator/extends.go [hunk 3]
```diff
@@ -249,6 +251,50 @@ func (g *generator) parseExtendFunc(fn *types.Func, opts *ParseExtendOptions) er
...
+			continue
+		case 1:
+			// name only, ignore empty inputs
+			namePattern = parts[0]
+			if namePattern == "" {
+				continue
+			}
+		}
+		// todo: handle other package path formats
+		pattern, err := regexp.Compile(namePattern)
```
Comment (jmattheis) [inline generator/extends.go:280]: This shouldn't need a pattern, as it is not possible to define multiple methods that are used in mapExtend.

**Review by jmattheis — CHANGES_REQUESTED** [review]

File: scenario/3_struct_mapping_extend.yml (hunk 2 above)
Comment (waltcow) [reply scenario/3_struct_mapping_extend.yml]: remove mapExtend parse in converter comment https://github.com/jmattheis/goverter/pull/25/commits/5b8e6df972144c269ed780cd1e060af8fe385d2b

File: generator/extends.go (hunk 3 above)
Comment (waltcow) [reply generator/extends.go]: remove method name pattern match in parse mapExtend process
 https://github.com/jmattheis/goverter/pull/25/commits/86abf4cb98b94958a1c3a14bc5b5a6fffe74d7cf

File: builder/struct.go (hunk 1 above)
Comment (waltcow) [reply builder/struct.go:45]: In my use case , pass single parameter into mapExtend function is needed,
currently, struct type or source type  as first parameter  are fully supported and tested

```golang
  func PluckAgeInt(source Input) int {
      return source.Age
  }

  func PlusScoreInt(source int) int {
      return source + 1
  }
```

Comment (jmattheis) [reply builder/struct.go:45]: Yes, but it is much easier to review if there are fewer features in one pull request. The use-case that was originally proposed in the ticket is not supported, and I don't think the current structure allows this to be cleanly implemented. (mapping extend method without parameter)

```go
// goverter:converter
type Converter interface {
    // goverter:mapExtend Age IntOne
    Convert(source Input) Output
}

type Input struct {
    Name string
}
type Output struct {
    Name string
    Age int
}
func IntOne() int {
    return 1
}
```

The BuildExtend method has code for returning errors and passing the converter instance to the extend method, this all is not tested.

Comment (waltcow) [reply builder/struct.go:45]: add  mapping extend method without parameter in commit https://github.com/jmattheis/goverter/pull/25/commits/e6688d753033217015307a6a5bf955833ce09cf1

Comment (waltcow) [reply builder/struct.go:45]: To make it simple,  I just remove support for returning errors and passing the converter instance to the extend method currently in commit https://github.com/jmattheis/goverter/pull/25/commits/4962136c2cb01ba62cabc19acc1242ba85a83ba6

**Comment by jmattheis** [comment]
I've rewritten most of the code, with a solution that didn't need changes on the `builder.Generator` interface, feel free to have a look at the code.

**Pull request closed without merging**